### ISSUE TITLE
Better error messages

### ISSUE THREAD

**Issue by mingodad** [ChrisHixon/chpeg#19]
Would be nice to have a call trace/stack accompanying the error messages.

Right now I'm trying to make a working Lua grammar and I'm getting this error message:
```
chpeg_backref lua.chpeg lpegrex.lua
chpeg_compile: Parse successful.
Grammar file compiled successfully. Parser returned: 0
Expected "}" in Table at:26:10 offset 594
Parse failed with result: 1
```
I know that the `input` is a valid Lua code (https://github.com/edubart/lpegrex/blob/main/lpegrex.lua).

Here is the backref grammar https://github.com/ChrisHixon/chpeg/issues/17#issuecomment-1144002456
And here the actual Lua grammar:
```
chunk <-
	 SHEBANG? SKIP Block ! . #~{ synerr(yy, "unexpected syntax"); }

SHEBANG <-
	 '#!' ( ! LINEBREAK . )* LINEBREAK?

SKIP <-
	 ( SPACE+ / COMMENT )*

Block <-
	 ( Label / Return / Break / Goto / Do / While / Repeat / If / ForNum / ForIn / FuncDef / FuncDecl / VarDecl / Assign / call / ( ';' SKIP ) )*

Label <-
	 '::' NAME '::' SKIP

Return <-
	 'return' SKIP exprlist?

Break <-
	 'break' SKIP

Goto <-
	 'goto' SKIP NAME

Do <-
	 'do' SKIP Block 'end' SKIP

While <-
	 'while' SKIP expr 'do' SKIP Block 'end' SKIP

Repeat <-
	 'repeat' SKIP Block 'until' SKIP expr

If <-
	 'if' SKIP expr 'then' SKIP Block ( 'elseif' SKIP expr 'then' SKIP Block )* ( 'else' SKIP Block )? 'end' SKIP

ForNum <-
	 'for' SKIP Id '=' SKIP expr ',' SKIP expr ( ',' SKIP expr )? 'do' SKIP Block 'end' SKIP

ForIn <-
	 'for' SKIP idlist 'in' SKIP exprlist 'do' SKIP Block 'end' SKIP

FuncDef <-
	 'function' SKIP funcname funcbody

FuncDecl <-
	 'local' SKIP 'function' SKIP Id funcbody

VarDecl <-
	 'local' SKIP iddecllist ( '=' SKIP exprlist )?

Assign <-
	 varlist '=' SKIP exprlist

call <-
	 exprprimary ( ( indexsuffix+ callsuffix ) / callsuffix )+

NAME <-
	 ! KEYWORD ( NAME_PREFIX NAME_SUFFIX? ) SKIP

exprlist <-
	 expr ( ',' SKIP expr )*

expr <-
	 expror

Id <-
	 NAME

idlist <-
	 Id ( ',' SKIP Id )*

funcname <-
	 Id DotIndex* ColonIndex?

funcbody <-
	 '(' SKIP funcargs ')' SKIP Block 'end' SKIP

iddecllist <-
	 IdDecl ( ',' SKIP IdDecl )*

varlist <-
	 var ( ',' SKIP var )*

Number <-
	 NUMBER SKIP

NUMBER <-
	 HEX_NUMBER
	/ DEC_NUMBER

String <-
	 STRING SKIP

STRING <-
	 STRING_SHRT
	/ STRING_LONG

Boolean <-
	 ( 'false' SKIP )
	/ ( 'true' SKIP )

Nil <-
	 'nil' SKIP

Varargs <-
	 '...' SKIP

IdDecl <-
	 NAME ( '<' SKIP NAME '>' SKIP )?

Function <-
	 'function' SKIP funcbody

Table <-
	 '{' SKIP ( field (( fieldsep field )* / fieldsep) )? '}' SKIP

field <-
	 Pair
	/ expr

fieldsep <-
	 ( ',' SKIP )
	/ ( ';' SKIP )

Paren <-
	 '(' SKIP expr ')' SKIP

Pair <-
	 ( '[' SKIP expr ']' SKIP '=' SKIP expr )
	/ ( NAME '=' SKIP expr )

Call <-
	 callargs

callargs <-
	 ( '(' SKIP ( expr ( ',' SKIP expr )* )? ')' SKIP )
	/ Table
	/ String

CallMethod <-
	 ':' SKIP NAME callargs

DotIndex <-
	 '.' SKIP NAME

ColonIndex <-
	 ':' SKIP NAME

KeyIndex <-
	 '[' SKIP expr ']' SKIP

indexsuffix <-
	 DotIndex
	/ KeyIndex

callsuffix <-
	 Call
	/ CallMethod

var <-
	 ( exprprimary ( ( callsuffix+ indexsuffix ) / indexsuffix )+ )
	/ Id

exprprimary <-
	 Id
	/ Paren

exprsuffixed <-
	 exprprimary ( indexsuffix / callsuffix )*

funcargs <-
	 ( ( Id ( ',' SKIP Id )* ( ',' SKIP Varargs )? ) / Varargs )?

opor <-
	 'or' SKIP exprand

exprand <-
	 exprcmp opand*

opand <-
	 'and' SKIP exprcmp

exprcmp <-
	 exprbor opcmp*

opcmp <-
	 ( ( '==' SKIP ) / ( '~=' SKIP ) / ( '<=' SKIP ) / ( '>=' SKIP ) / ( '<' SKIP ) / ( '>' SKIP ) ) exprbor

exprbor <-
	 exprbxor opbor*

opbor <-
	 '|' SKIP exprbxor

exprbxor <-
	 exprband opbxor*

opbxor <-
	 '~' SKIP exprband

exprband <-
	 exprbshift opband*

opband <-
	 '&' SKIP exprbshift

exprbshift <-
	 exprconcat opbshift*

opbshift <-
	 ( ( '<<' SKIP ) / ( '>>' SKIP ) ) exprconcat

exprconcat <-
	 exprarit opconcat*

opconcat <-
	 '..' SKIP exprconcat

oparit <-
	 ( ( '+' SKIP ) / ( '-' SKIP ) ) exprfact

exprfact <-
	 exprunary opfact*

opfact <-
	 ( ( '*' SKIP ) / ( '//' SKIP ) / ( '/' SKIP ) / ( '%' SKIP ) ) exprunary

exprunary <-
	 opunary
	/ exprpow

oppow <-
	 '^' SKIP exprunary

opunary <-
	 ( ( 'not' SKIP ) / ( '#' SKIP ) / ( '-' SKIP ) / ( '~' SKIP ) ) exprunary

expror <-
	 exprand opor*

exprarit <-
	 exprfact oparit*

exprpow <-
	 exprsimple oppow*

exprsimple <-
	 Nil
	/ Boolean
	/ Number
	/ String
	/ Varargs
	/ Function
	/ Table
	/ exprsuffixed

STRING_SHRT <-
	 ["'] ( ESCAPE_SEQ / ( ! ( $1 / LINEBREAK ) . ) )* $1

STRING_LONG <-
	 '[' '='* '[' LINEBREAK? ( ! (']' $2 ']') . )* ']' $2 ']'

ESCAPE_SEQ <-
	 '\\' ESCAPE

LINEBREAK <-
	 '\r\n'
	/ '\n\r'
	/ '\n'
	/ '\r'

ESCAPE <-
	 [\'"]
	/ ( 'n' / 't' / 'r' / 'a' / 'b' / 'v' / 'f' )
	/ ( 'x' HEX_DIGIT HEX_DIGIT )
	/ ( 'u' '{' HEX_DIGIT HEX_DIGIT+ '}' )
	/ ( 'z' SPACE* )
	/ (DEC_DIGIT DEC_DIGIT !DEC_DIGIT / [012] DEC_DIGIT DEC_DIGIT)
	/ LINEBREAK

HEX_DIGIT <-
	 [0-9a-fA-F]

SPACE <-
	 ' '
	/ '\t'
	/ LINEBREAK

HEX_NUMBER <-
	 '0' [xX] HEX_PREFIX ( [pP] EXP_DIGITS )?

DEC_NUMBER <-
	 DEC_PREFIX ( [eE] EXP_DIGITS )?

HEX_PREFIX <-
	 ( HEX_DIGIT+ ( '.' HEX_DIGIT* )? )
	/ ( '.' HEX_DIGIT+ )

EXP_DIGITS <-
	 [+-]? DEC_DIGIT+

DEC_PREFIX <-
	 ( DEC_DIGIT+ ( '.' DEC_DIGIT* )? )
	/ ( '.' DEC_DIGIT+ )

DEC_DIGIT <-
	 [0-9]

COMMENT <-
	 '--' ( STRING_LONG / COMMENT_SHRT )

COMMENT_SHRT <-
	 ( ! LINEBREAK . )*

KEYWORD <-
	 ('and'
	/ 'break'
	/ 'do'
	/ 'elseif'
	/ 'else'
	/ 'end'
	/ 'false'
	/ 'for'
	/ 'function'
	/ 'goto'
	/ 'if'
	/ 'in'
	/ 'local'
	/ 'nil'
	/ 'not'
	/ 'or'
	/ 'repeat'
	/ 'return'
	/ 'then'
	/ 'true'
	/ 'until'
	/ 'while') !NAME_PREFIX

NAME_PREFIX <-
	 [_a-zA-Z]

NAME_SUFFIX <-
	 [_a-zA-Z0-9]+

EXTRA_TOKENS <-
	 '[[' '[=' '--'
```

**Comment by mingodad** [comment]
I think that having something like https://github.com/yhirose/cpp-peglib#cut-operator or https://github.com/sqmedeiros/lpeglabel/ or to force error/prevent backtracking would allow better error messages.

Like after several tests I found that the error message `Expected "}" in Table at:26:10 offset 594` is due to a `STRING_SHRT` not matching and with something like:
```
STRING_SHRT <-
	 ["'] ~{"string match error"} ( ESCAPE_SEQ / ( ! ( $1 / LINEBREAK ) . ) )* $1
```
We could have an error message like:
```
string match error at:26:10 offset 594
```

**Comment by ChrisHixon** [comment]
Displaying useful error messages automatically in PEG is not a simple task. There is not a point where we can just detect an error, exit, and dump a stack trace. A parse fails only after backtracking and trying everything possible. The most common strategy, which chpeg uses, is to track the last (most current) mismatch/failure that occurs as far as we can get into the input.

This experiment tracks and displays ALL errors at the farthest point we've reached in the input: https://github.com/ChrisHixon/chpeg/tree/errors-experiment

Example:
```
% util/chpeg-trace -v -g bugs/c99-mouse-files/c99-mouse.peg -P 'void f(int g'       
Compiling grammar file: 'bugs/c99-mouse-files/c99-mouse.peg'
Grammar file 'bugs/c99-mouse-files/c99-mouse.peg' compiled successfully. Parser returned: 0
Parsing string: "void f(int g"
Parse failed with result: 1
  Expected:                     [a-z] in Identifier at offset 12
  Expected:                     [A-Z] in Identifier at offset 12
  Expected:                     [0-9] in Identifier at offset 12
  Expected:                     [_] in Identifier at offset 12
  Expected:                     "\\u" in Identifier at offset 12
  Expected:                     "\\U" in Identifier at offset 12
  Expected:                     UniversalCharacter in DirectDeclarator at offset 12
  Expected:                     IdChar in DirectDeclarator at offset 12
  Expected:                     [ \n\r\t] in Identifier at offset 12
  Expected:                     WhiteSpace in DirectDeclarator at offset 12
  Expected:                     "/*" in Identifier at offset 12
  Expected:                     LongComment in DirectDeclarator at offset 12
  Expected:                     "//" in Identifier at offset 12
  Expected:                     LineComment in DirectDeclarator at offset 12
  Expected:                     "#" in Identifier at offset 12
  Expected:                     Pragma in DirectDeclarator at offset 12
  Expected:                     "__extension__" in Identifier at offset 12
  Expected:                     EXTENSION in DirectDeclarator at offset 12
  Expected:                     "[" in LBRK at offset 12
  Expected:                     LBRK in DirectDeclarator at offset 12
  Expected:                     "[" in LBRK at offset 12
  Expected:                     LBRK in DirectDeclarator at offset 12
  Expected:                     "[" in LBRK at offset 12
  Expected:                     LBRK in DirectDeclarator at offset 12
  Expected:                     "[" in LBRK at offset 12
  Expected:                     LBRK in DirectDeclarator at offset 12
  Expected:                     "(" in LPAR at offset 12
  Expected:                     LPAR in DirectDeclarator at offset 12
  Expected:                     "(" in LPAR at offset 12
  Expected:                     LPAR in DirectDeclarator at offset 12
  Expected:                       "__attribute__" in ATTRIBUTE at offset 12
  Expected:                       ATTRIBUTE in attribute at offset 12
  Expected:                     attribute in ExtensionSpecifier at offset 12
  Expected:                       "__asm__" in ASM at offset 12
  Expected:                       "__asm" in ASM at offset 12
  Expected:                       ASM in asm at offset 12
  Expected:                     asm in ExtensionSpecifier at offset 12
  Expected:                   ExtensionSpecifier in Declarator at offset 12
  Expected:               "," in COMMA at offset 12
  Expected:               COMMA in ParameterList at offset 12
  Expected:             "," in COMMA at offset 12
  Expected:             COMMA in ParameterTypeList at offset 12
  Expected:           ")" in RPAR at offset 12
  Expected:           RPAR in DirectDeclarator at offset 12
  Expected:                         [a-z] in Identifier at offset 12
  Expected:                         [A-Z] in Identifier at offset 12
  Expected:                         [0-9] in Identifier at offset 12
  Expected:                         [_] in Identifier at offset 12
  Expected:                         "\\u" in Identifier at offset 12
  Expected:                         "\\U" in Identifier at offset 12
  Expected:                         UniversalCharacter in DirectDeclarator at offset 12
  Expected:                         IdChar in DirectDeclarator at offset 12
  Expected:                         [ \n\r\t] in Identifier at offset 12
  Expected:                         WhiteSpace in DirectDeclarator at offset 12
  Expected:                         "/*" in Identifier at offset 12
  Expected:                         LongComment in DirectDeclarator at offset 12
  Expected:                         "//" in Identifier at offset 12
  Expected:                         LineComment in DirectDeclarator at offset 12
  Expected:                         "#" in Identifier at offset 12
  Expected:                         Pragma in DirectDeclarator at offset 12
  Expected:                         "__extension__" in Identifier at offset 12
  Expected:                         EXTENSION in DirectDeclarator at offset 12
  Expected:                         "[" in LBRK at offset 12
  Expected:                         LBRK in DirectDeclarator at offset 12
  Expected:                         "[" in LBRK at offset 12
  Expected:                         LBRK in DirectDeclarator at offset 12
  Expected:                         "[" in LBRK at offset 12
  Expected:                         LBRK in DirectDeclarator at offset 12
  Expected:                         "[" in LBRK at offset 12
  Expected:                         LBRK in DirectDeclarator at offset 12
  Expected:                         "(" in LPAR at offset 12
  Expected:                         LPAR in DirectDeclarator at offset 12
  Expected:                         "(" in LPAR at offset 12
  Expected:                         LPAR in DirectDeclarator at offset 12
  Expected:                           "__attribute__" in ATTRIBUTE at offset 12
  Expected:                           ATTRIBUTE in attribute at offset 12
  Expected:                         attribute in ExtensionSpecifier at offset 12
  Expected:                           "__asm__" in ASM at offset 12
  Expected:                           "__asm" in ASM at offset 12
  Expected:                           ASM in asm at offset 12
  Expected:                         asm in ExtensionSpecifier at offset 12
  Expected:                       ExtensionSpecifier in Declarator at offset 12
  Expected:                   "," in COMMA at offset 12
  Expected:                   COMMA in ParameterList at offset 12
  Expected:                 "," in COMMA at offset 12
  Expected:                 COMMA in ParameterTypeList at offset 12
  Expected:               ")" in RPAR at offset 12
  Expected:               RPAR in DirectDeclarator at offset 12
```
It might somewhat resemble a stack trace, but it is simply logging the tree stack depth (used for indenting here), the rule we're in (the right hand part 'in XXXXXX') and pc (instruction address) which is used later to get the rest of the info.
The last line of this is what would normally be displayed as an error.

I did this hoping to come up with a way to look back at this 'error log' and create a better error message. But not sure this is making it any easier!

**Comment by ChrisHixon** [comment]
One thing I'd like to try is not tracking anything optional, like mismatches while we're 'inside' `*` and `?` operations, and see how that affects things. Not matching an optional thing causes backtracking but is not an error location. "Error: expecting optional whitespace" is silly. I think I'll call this a bug in current tracking.

**Comment by ChrisHixon** [comment]
I think I was onto something like this once, there are disabled remnants in the code. See things guarded by `ERROR_REPEAT_INHIBIT`. Error tracking has to be locked and disabled when entering all optional states... which I think can be boiled down to `*`, '?' and anything after the first `+` match.

**Comment by mingodad** [comment]
And what about the `cut` operator or `label` like in [le](https://github.com/sqmedeiros/lpeglabel/) ?

**Comment by ChrisHixon** [comment]
I will have to look into things like that, but this 'optional' issue is certainly a flaw in the current code.

**Comment by mingodad** [comment]
Is to much to ask for a mini tutorial/documentation of the `VM` ?
Like you mentioned the `locked and disabled` what that mean in the `VM` I can see there is a variable `locked` in the code.

**Comment by ChrisHixon** [comment]
`locked` is a state variable that (when non-zero) means parse tree building is locked, meaning currently disabled. It's used when it's inside an ignored `{I}` or leaf `{L}` rule where we don't want parse tree nodes created.
I need to do something similar to inhibit error tracking when processing optional sequences, and it appears to be in place already, err_locked, with all the code guarded by ERROR_REPEAT_INHIBIT. I had a comment that this idea was flawed, do not enable, so maybe it's not quite right, but so far going through it it looks like what is needed.
I've been trying to add more comments in the VM as I work on it. Unfortunately, it's getting really cluttered up now with optional things: tracing, trim, packrat, etc.

**Comment by mingodad** [comment]
Can we try online pair programming ?

**Comment by ChrisHixon** [comment]
Oh I see, I wasn't quite thinking it all through. It completely messes up error tracking disabling during optional repeats.

Using this calc example, as soon as you have a [+-/*] error tracking is disabled. Not good!

```
Expr <- _ Sum EOF

Sum <- Product (PLUS_MINUS _ Product )*
Product <- Value (MUL_DIV _ Value )*
Value <- NUMBER _ / ParenExpr
ParenExpr <- '(' _ Sum ')' _

PLUS_MINUS <- [+-]
MUL_DIV <- [*/]
NUMBER <- "0" / [1-9][0-9]*
_ {I} <- [ \t\r\n]*

EOF <- !.
```

**Comment by ChrisHixon** [comment]
From another issue:

> > Yes it's working now ! Thank you so much ! Maybe it's a good idea to extend the error message with a description of the error, otherwise we end up in auto mode and ignore anything `result: \d+` because it's not relevant most of the time. Something like:
> > ```
> > Parse failed with result: 3 -> Stack overflow
> > ```
> 
> Yes that's a good idea. Also, printing the tracked parse errors makes no sense unless it's a normal parse failure ( CHPEG_ERR_PARSE_FAILED = 1)... they will be nonsensical like what we saw here.
